File: src/structural.ts
interface HasSsn {
    ssn: string
}

interface Credit {
    income: number
    age: number
}

interface HasCredit {
    credit: Credit
}

interface HasNoCredit {
    credit: undefined
}

function hasSsn(v: any): v is HasSsn {
    return v.ssn != null
}

function creditLookup<P>(person: P | P & HasSsn): P & HasNoCredit | P & HasSsn & HasCredit {
    if (hasSsn(person)) {
        return Object.assign({ credit: { income: 2000, age: 42 } }, person)
    }
    return Object.assign({ credit: undefined }, person)
}

const data = [
    {ssn: "1234"},
    {name:  "foo"},
    {ssn: "1234", name: "foo"}]

console.log(data.map(creditLookup))

File: src/structural.clj
(defn credit-lookup [{:keys [ssn] :as person}]
  (assoc person :credit (when ssn {:income 2000 :age 42})))

(map credit-lookup
     [{:ssn "1234"}
      {:name "foo"}
      {:ssn "1234" :name "foo"}])
